x/ecocredit/genesis.go

```diff
@@ -60,9 +61,61 @@ func (s *GenesisState) Validate() error {
 		return err
 	}
 
+	// run each params validation method
+	if err := s.Params.Validate(); err != nil {
+		return err
+	}
+
+	// check that the CreditTypes in the ClassInfo slice all exist in params.CreditTypes
+	if err := validateClassInfoTypes(s.Params.CreditTypes, s.ClassInfo); err != nil {
+		return err
+	}
+
 	return nil
 }
 
+func validateClassInfoTypes(creditTypes []*CreditType, classInfos []*ClassInfo) error {
+	typeMap := make(map[string]CreditType, len(creditTypes))
+
+	// convert to map for easier lookups
+	for _, cType := range creditTypes {
+		typeMap[cType.Abbreviation] = *cType
+	}
+
+	for _, cInfo := range classInfos {
+		// fetch param via abbreviation
+		cType, ok := typeMap[cInfo.CreditType.Abbreviation]
+
+		// if its not found, its an invalid credit type
+		if !ok {
+			return fmt.Errorf("invalid credit type abbreviation: %s", cInfo.CreditType.Abbreviation)
+		}
+
+		// check that the other fields are equal
+		if !areCreditTypesEqual(cType, *cInfo.CreditType) {
+			return fmt.Errorf("credit type %s does not match param type %s", cInfo.CreditType.Name, cType.Name)
+		}
+	}
+	return nil
+}
+
+// checks that fields are equal between two credits.
+func areCreditTypesEqual(t1, t2 CreditType) bool {
```

Comment:
Might be nice to use a switch statement here.
```
switch {
case t1.Abbreviation != t2.Abbreviation:
case t1.Name != t2.Name:
case t1.Precision != t2.Precision:
case t1.Unit != t2.Unit:
    return false
default:
    return true
}
```

Reply:
Out of interest, why doesn't `t1 == t2` work? Could you document that in this function?

Reply:
oops, looks like you can just do `t1 == t2` updated here https://github.com/regen-network/regen-ledger/pull/511/commits/79cbd50fdd4ff8c8666f1f4d7d75a2852b0c3613#diff-e57c839aca633b049a6d8f6706c7b5c53322c1e7cc17157178da38461789043cR95